nrf_rpc/include/nrf_rpc.h

```diff
@@ -0,0 +1,444 @@
+			      "evt_" NRF_RPC_STRINGIFY(_name));		       \
+	NRF_RPC_AUTO_ARR_ITEM(const struct nrf_rpc_group, _name, "grp",	       \
+			       NRF_RPC_STRINGIFY(_name)) = {		       \
+		.group_id = (_id),					       \
+		.cmd_array = &NRF_RPC_CONCAT(_name, _cmd_array),	       \
+		.evt_array = &NRF_RPC_CONCAT(_name, _evt_array),	       \
+	}
+
+
+/** @brief Register a command decoder.
+ *
+ * @param _group   Group that the decoder will belong to.
+ * @param _name    Name of the decoder.
+ * @param _cmd     Command id. Can be from 0 to 255.
+ * @param _handler Handler function of type @a nrf_rpc_handler_t.
+ */
+#define NRF_RPC_CMD_DECODER(_group, _name, _cmd, _handler, _data)	       \
+	NRF_RPC_STATIC_ASSERT(_cmd <= 0xFE, "Command out of range");	       \
+	NRF_RPC_AUTO_ARR_ITEM(const struct nrf_rpc_decoder,		       \
+			       NRF_RPC_CONCAT(_name, _cmd_dec),		       \
+			       "cmd_" NRF_RPC_STRINGIFY(_group),	       \
+			       NRF_RPC_STRINGIFY(_name)) = {		       \
+		.id = _cmd,						       \
+		.handler = _handler,					       \
+		.handler_data = _data,					       \
+	}
+
+
+/** @brief Register an event decoder.
+ *
+ * @param _group   Group that the decoder will belong to.
+ * @param _name    Name of the decoder.
+ * @param _evt     Event id. Can be from 0 to 255.
+ * @param _handler Handler function of type @a nrf_rpc_handler_t.
+ */
+#define NRF_RPC_EVT_DECODER(_group, _name, _evt, _handler, _data)	       \
+	NRF_RPC_STATIC_ASSERT(_evt <= 0xFE, "Event out of range");	       \
+	NRF_RPC_AUTO_ARR_ITEM(const struct nrf_rpc_decoder,		       \
+			       NRF_RPC_CONCAT(_name, _evt_dec),		       \
+			       "evt_" NRF_RPC_STRINGIFY(_group),	       \
+			       NRF_RPC_STRINGIFY(_name)) = {		       \
+		.id = _evt,						       \
+		.handler = _handler,					       \
+		.handler_data = _data,					       \
+	}
+
+
+/** @brief Allocates resources for a new command.
+ *
+ * Macro may allocate some variables on stack, so it should be used at top level
+ * of a function.
+ *
+ * @param      _group  Group that command belongs to.
+ * @param[out] _ctx    Variable of type `nrf_rpc_cmd_ctx_t` that will hold
+ *                     allocated resources.
+ * @param[out] _packet Variable of type `uint8_t *` that will hold pointer to
+ *                     newly allocated packet buffer.
+ * @param      _len    Requested length of the packet.
+ * @param __VA_ARGS__  Code that will be executed in case of allocation failure.
+ *                     This can be e.g. return or goto statement.
+ */
+#define NRF_RPC_CMD_ALLOC(_ctx, _group, _packet, _len, ...)		       \
+	(_ctx) = _nrf_rpc_cmd_prep((_group));				       \
+	nrf_rpc_tr_alloc_tx_buf((_ctx), &(_packet),			       \
+				_NRF_RPC_HEADER_SIZE + (_len));		       \
+	if (nrf_rpc_tr_alloc_failed((_packet))) {			       \
+		_nrf_rpc_cmd_alloc_error((_ctx));			       \
+		__VA_ARGS__;						       \
+	}								       \
+	(_packet) += _NRF_RPC_HEADER_SIZE
```

Comment:
I agree that `__VA_ARGS__` solution is not very elegant, but I decided to do it this way, because it is simple and fulfills all the assumptions.

I don't understand your suggestion. I don't see how `NRF_RPC_CMD_DEFINE` macro is related to this issue. Can you write how exactly `NRF_RPC_CMD_ALLOC` macro should looks like in your opinion. Remember about following:
* There may be different implementations of the transport layer. Currently we have only rpmsg, but it will change in the future.
* Transport layer decides how and where the buffer is allocated.
   * It may be on stack or not.
   * It may be allocated directly in the shared memory to implement zero-copying.
   * It may be in some buffers intended to be send directly via UART.

Reply:
I can agree that non-stack allocation can block until enough free space.

Reply:
I changed the allocation based on suggestions from above comments, so it will not return an error, but it will wait. This reduced error handling in this context a lot.